File: 250119/39_combination_sum.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    std::vector<std::vector<int>> combinationSum(std::vector<int>& candidates, int target) {
        std::vector<std::vector<int>> result;
        std::vector<int> vec;
        helper(candidates, vec, result, 0, 0, target);
        return result;
    }

    void helper(std::vector<int>& candidates, std::vector<int>& vec, std::vector<std::vector<int>> &result, int index, int total, int target) {
        if (total == target) {
            result.push_back(vec);
            return;
        }

        for (int i = index; i < candidates.size(); i++) {
            int val = candidates[i];
            if (total + val <= target) {
                vec.push_back(val);
                helper(candidates, vec, result, i, total + val, target);
                vec.pop_back();
            }
        }

    }

    void printResult(const std::vector<std::vector<int>> result) {
        std::cout << "Unique combinations: \n";
        for (const auto& res : result) {
            std::cout << "[";
            for (int num : res) std:: cout << num << " ";
            std::cout << "]\n";
        }
    }
};

int main() {
    Solution sol;
    std::vector<int> candidates = {2,3,6,7};
    int target = 7;

    std::vector<std::vector<int>> result = sol.combinationSum(candidates, target);
    sol.printResult(result);
    
    return 0;
}

File: 250119/39_combination_sum.py
# 39. Combination Sum

from typing import List

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []
        self.helper([], candidates, 0, 0, target, result)
        return result
    
    def helper(self, list_, candidates, index, total, target, result):
        if total == target:
            result.append(list_)
            return
        
        for i in range(index, len(candidates)):
            n = candidates[i]
            if total + n <= target:
                self.helper(list_ + [n], candidates, i, total + n, target, result)

if __name__ == "__main__":
    sol = Solution()

    candidates = [2,3,6,7]
    target = 7

    result = sol.combinationSum(candidates, target)
    print("Unique combinations of candidates: ", result)
